src/datasets/utils/extract.py

```diff
@@ -0,0 +1,186 @@
+import gzip
+import lzma
+import os
+import shutil
+import struct
+import tarfile
+from zipfile import ZipFile
+from zipfile import is_zipfile as _is_zipfile
+
+from datasets import config
+from datasets.utils.filelock import FileLock
+
+
+class ExtractManager:
+    def __init__(self, cache_dir=None):
+        self.extract_dir = (
+            os.path.join(cache_dir, config.EXTRACTED_DATASETS_DIR) if cache_dir else config.EXTRACTED_DATASETS_PATH
+        )
+        self.extractor = Extractor
+
+    def _get_output_path(self, path):
+        from datasets.utils.file_utils import hash_url_to_filename
+
+        # Path where we extract compressed archives
+        # We extract in the cache dir, and get the extracted path name by hashing the original path"
+        abs_path = os.path.abspath(path)
+        return os.path.join(self.extract_dir, hash_url_to_filename(abs_path))
+
+    def _do_extract(self, output_path, force_extract):
+        return force_extract or (
+            not os.path.isfile(output_path) and not (os.path.isdir(output_path) and os.listdir(output_path))
+        )
+
+    def extract(self, input_path, force_extract=False):
+        is_extractable, extractor = self.extractor.is_extractable(input_path, return_extractor=True)
+        if not is_extractable:
+            return input_path
+        output_path = self._get_output_path(input_path)
+        if self._do_extract(output_path, force_extract):
+            try:
+                self.extractor.extract(input_path, output_path, extractor=extractor)
+            except Exception:
```

Comment:
Maybe only catch only the exception that happens when the archive formatted is not identified ? Which errors would you like to catch here ?

Otherwise this could catch unexpected errors.

You can also add the message of the original error to the `EnvironmentError` as additional information

Reply:
You are right: in the original code there was no exception catch. 

Indeed, in the original code the raise of `EnvironmentError` was dead code: 
- previously if none of the if conditions was met there was a return (`return output_path`), 
  - See https://github.com/huggingface/datasets/blob/d7a72234436673d6394bc67cccb5489aacf5f981/src/datasets/utils/file_utils.py#L305-L313
- and this part of the code was in the else clause after all of the if conditions repeated again: the else clause was not possible to be executed because of the previous condition I mentioned above
  - See: https://github.com/huggingface/datasets/blob/d7a72234436673d6394bc67cccb5489aacf5f981/src/datasets/utils/file_utils.py#L338-L369

I would suggest just remove this exception raise and the corresponding try and the exception catch.